common/rfb/CMsgWriter.cxx

```diff
@@ -176,15 +176,37 @@ void CMsgWriter::writeKeyEvent(uint32_t keysym, uint32_t keycode, bool down)
 void CMsgWriter::writePointerEvent(const Point& pos, uint16_t buttonMask)
 {
   Point p(pos);
+  bool extendedMouseButtons;
+
   if (p.x < 0) p.x = 0;
   if (p.y < 0) p.y = 0;
   if (p.x >= server->width()) p.x = server->width() - 1;
   if (p.y >= server->height()) p.y = server->height() - 1;
 
+  // Only send extended pointerEvent message when needed
+  extendedMouseButtons = buttonMask & 0x180;
+
   startMsg(msgTypePointerEvent);
-  os->writeU8(buttonMask);
-  os->writeU16(p.x);
-  os->writeU16(p.y);
+  if (server->supportsExtendedMouseButtons && extendedMouseButtons) {
+    int higherBits;
+    int lowerBits;
+
+    higherBits = (buttonMask  >> 7) & 0xff;
+    lowerBits = buttonMask & 0x7f;
+    lowerBits |= 0x80; // Set marker bit to 1
+
+    higherBits &= 0x03; // Clear reserved bits
+
+    os->writeU8(lowerBits);
+    os->writeU16(p.x);
+    os->writeU16(p.y);
+    os->writeU8(higherBits);
+  } else {
+    buttonMask &= 0x7f; // Set marker bit to 0
```

Comment:
We probably need a comment why this is needed, or someone might easily remove it.